src/compute/arithmetics/decimals/add.rs

```diff
@@ -0,0 +1,440 @@
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Defines the addition arithmetic kernels for Decimal `PrimitiveArrays`.
+
+use crate::{
+    array::{Array, PrimitiveArray},
+    buffer::Buffer,
+};
+use crate::{
+    datatypes::DataType,
+    error::{ArrowError, Result},
+};
+
+use super::{adjusted_precision_scale, max_value, number_digits};
+use crate::compute::arity::{binary, try_binary};
+use crate::compute::utils::combine_validities;
+
+/// Adds two decimal primitive arrays with the same precision and scale. If the
+/// precision and scale is different, then an InvalidArgumentError is returned.
+/// This function panics if the added numbers result in a number larger than
+/// the possible number for the selected precision.
+pub fn add(lhs: &PrimitiveArray<i128>, rhs: &PrimitiveArray<i128>) -> Result<PrimitiveArray<i128>> {
+    // Matching on both data types from both arrays
+    // This match will be true only when precision and scale from both
+    // arrays are the same, otherwise it will return and ArrowError
+    match (lhs.data_type(), rhs.data_type()) {
+        (DataType::Decimal(lhs_p, lhs_s), DataType::Decimal(rhs_p, rhs_s)) => {
+            if lhs_p == rhs_p && lhs_s == rhs_s {
+                // Closure for the binary operation. This closure will panic if
+                // the sum of the values is larger than the max value possible
+                // for the decimal precision
+                let op = move |a, b| {
+                    let res: i128 = a + b;
+
+                    if res.abs() > max_value(*lhs_p) {
+                        panic!("Overflow in addition presented for precision {}", lhs_p);
+                    }
+
+                    res
+                };
+
+                binary(lhs, rhs, op)
+            } else {
+                return Err(ArrowError::InvalidArgumentError(
+                    "Arrays must have the same precision and scale".to_string(),
+                ));
+            }
+        }
+        _ => {
+            return Err(ArrowError::InvalidArgumentError(
+                "Incorrect data type for the array".to_string(),
+            ));
+        }
+    }
+}
+
+/// Saturated addition of two decimal primitive arrays with the same precision
+/// and scale. If the precision and scale is different, then an
+/// InvalidArgumentError is returned. If the result from the sum is larger than
+/// the possible number with the selected precision then the resulted number in
+/// the arrow array is the maximum number for the selected precision.
+pub fn saturating_add(
```

Comment:
The same argument applies here: this could also work for non-i128 primitive types. The datatype defines behavior.